Data/ByteString/Builder/RealFloat.hs

```diff
@@ -0,0 +1,227 @@
+
+module Data.ByteString.Builder.RealFloat
+  ( FloatFormat
+  , floatDec
+  , doubleDec
+  , fixed
+  , fixedDefaultPrecision
+  , scientific
+  , generic
+  , formatFloat
+  , formatDouble
+  ) where
+
+import Data.ByteString.Builder.Internal (Builder)
+import qualified Data.ByteString.Builder.RealFloat.Internal as R
+import qualified Data.ByteString.Builder.RealFloat.F2S as RF
+import qualified Data.ByteString.Builder.RealFloat.D2S as RD
+import qualified Data.ByteString.Builder.Prim as BP
+import GHC.Float (roundTo)
+import GHC.Word (Word64)
+import GHC.Show (intToDigit)
+
+-- | Returns a rendered Float. Matches `show` in displaying in fixed or
+-- scientific notation
+{-# INLINABLE floatDec #-}
+floatDec :: Float -> Builder
+floatDec = formatFloat generic
+
+-- | Returns a rendered Double. Matches `show` in displaying in fixed or
+-- scientific notation
+{-# INLINABLE doubleDec #-}
+doubleDec :: Double -> Builder
+doubleDec = formatDouble generic
+
+-- | An opaque wrapper around `FloatFormat'` and an optional precision argument
+-- used by `formatFloat` and `formatDouble`.
+data FloatFormat = MkFloatFormat
+  { floatFormat' :: FloatFormat'
+  , precision :: Maybe Int
+  }
+
+-- | Standard notation with `n` decimal places
+fixed :: Int -> FloatFormat
+fixed n = MkFloatFormat FFixed (Just n)
+
+-- | Standard notation with the default precision (number of decimal places)
+fixedDefaultPrecision :: FloatFormat
+fixedDefaultPrecision = MkFloatFormat FFixed Nothing
+
+-- | Scientific notation with default precision
+scientific :: FloatFormat
+scientific = MkFloatFormat FScientific Nothing
+
+-- | Standard or scientific notation depending on the exponent. Matches `show`
+generic :: FloatFormat
+generic = MkFloatFormat FGeneric Nothing
+
+-- | ByteString float-to-string format
+data FloatFormat'
+  = FScientific     -- ^ scientific notation
+  | FFixed          -- ^ standard notation with `Maybe Int` digits after the decimal
+  | FGeneric        -- ^ dispatches to scientific or standard notation based on the exponent
+  deriving Show
+
+-- TODO: support precision argument for FGeneric and FScientific
+-- | Returns a rendered Float. Matches the API of `formatRealFloat` but does
+-- not currently handle the precision argument in scientific notation.
+--
+-- The precision argument is used to truncate (or extend with 0s) the
+-- 'shortest' rendered Float. A precision of 'Nothing' does no such
+-- modifications and will return as many decimal places as the representation
+-- demands.
+--
+-- e.g
+--
+-- >>> formatFloat (fixed 1) 1.2345e-2
+-- "0.0"
+-- >>> formatFloat (fixed 5) 1.2345e-2
+-- "0.01234"
+-- >>> formatFloat (fixed 10) 1.2345e-2
+-- "0.0123450000"
+-- >>> formatFloat fixedDefaultPrecision 1.2345e-2
+-- "0.01234"
+-- >>> formatFloat scientific 12.345
+-- "1.2345e1"
+-- >>> formatFloat generic 12.345
+-- "12.345"
+{-# INLINABLE formatFloat #-}
+formatFloat :: FloatFormat -> Float -> Builder
+formatFloat (MkFloatFormat fmt prec) f =
```

Comment:
Yes, it will be better for inlining this way.

Reply:
I updated this. Might there be problems with assuming https://github.com/haskell/bytestring/pull/365/files#diff-3016af887ac9c7d42c56d7358eacf8168698724938c79cf8239cdd4957eaccdaR114 is lazy?

Reply:
Can you clarify what kind of problems you're concerned about? The easiest thing might be to take a look at the Core…

Reply:
Just that running `f2Intermediate` is unnecessary work in case of `FScientific`. I took a look at the `dump-simpl` output but it wasn't obvious to me what I should be looking for...